File: gauss.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func gaussElimination(matrix [][]float64) ([]float64, error) {
	n := len(matrix)
	solution := make([]float64, n)

	
	m := make([][]float64, n)
	for i := range matrix {
		m[i] = make([]float64, len(matrix[i]))
		copy(m[i], matrix[i])
	}

	for i := 0; i < n; i++ {
	
		maxRow := i
		for j := i; j < n; j++ {
			if math.Abs(m[j][i]) > math.Abs(m[maxRow][i]) {
				maxRow = j
			}
		}
		m[i], m[maxRow] = m[maxRow], m[i]

	
		if math.Abs(m[i][i]) < 1e-10 {
			return nil, fmt.Errorf("matriz singular, sem solução única")
		}

		for j := i + 1; j < n; j++ {
			factor := m[j][i] / m[i][i]
			for k := i; k <= n; k++ {
				m[j][k] -= factor * m[i][k]
			}
		}
	}

	for i := n - 1; i >= 0; i-- {
		solution[i] = m[i][n]
		for j := i + 1; j < n; j++ {
			solution[i] -= m[i][j] * solution[j]
		}
		solution[i] /= m[i][i]
	}

	return solution, nil
}

func main() {
	reader := bufio.NewReader(os.Stdin)

	fmt.Print("Número de variáveis: ")
	nStr, _ := reader.ReadString('\n')
	n, err := strconv.Atoi(strings.TrimSpace(nStr))
	if err != nil || n <= 0 {
		fmt.Println("Número inválido de variáveis")
		return
	}

	matrix := make([][]float64, n)
	fmt.Println("Digite a matriz aumentada (linha por linha):")

	for i := 0; i < n; i++ {
		line, _ := reader.ReadString('\n')
		fields := strings.Fields(strings.TrimSpace(line))
		
		if len(fields) != n+1 {
			fmt.Println("Número inválido de elementos na linha")
			return
		}
		
		row := make([]float64, n+1)
		for j, field := range fields {
			val, err := strconv.ParseFloat(field, 64)
			if err != nil {
				fmt.Println("Valor não numérico detectado")
				return
			}
			row[j] = val
		}
		matrix[i] = row
	}

	solution, err := gaussElimination(matrix)
	if err != nil {
		fmt.Println("Erro:", err)
		return
	}

	fmt.Println("\nSolução:")
	for i, val := range solution {
		fmt.Printf("x[%d] = %.6f\n", i, val)
	}
}

File: gauss.rs
use std::error::Error;
use std::io;
use std::io::Write;

type Matrix = Vec<Vec<f64>>;

fn eliminacao_gauss(matriz: &mut Matrix) -> Result<Vec<f64>, Box<dyn Error>> {
    let n = matriz.len();
    let mut solucao = vec![0.0; n];

    // Fase de eliminação progressiva
    for i in 0..n {
        // Pivotamento parcial
        let mut max_linha = i;
        for j in i..n {
            if matriz[j][i].abs() > matriz[max_linha][i].abs() {
                max_linha = j;
            }
        }
        matriz.swap(i, max_linha);

        // Verifica matriz singular
        if matriz[i][i].abs() < 1e-10 {
            return Err("Matriz singular, sem solução única".into());
        }

        // Eliminação
        for j in (i + 1)..n {
            let fator = matriz[j][i] / matriz[i][i];
            for k in i..(n + 1) {
                matriz[j][k] -= fator * matriz[i][k];
            }
        }
    }

    // Substituição regressiva
    for i in (0..n).rev() {
        solucao[i] = matriz[i][n];
        for j in (i + 1)..n {
            solucao[i] -= matriz[i][j] * solucao[j];
        }
        solucao[i] /= matriz[i][i];
    }

    Ok(solucao)
}

fn main() -> Result<(), Box<dyn Error>> {
    print!("Número de variáveis: ");
    io::stdout().flush()?;
    
    let mut input = String::new();
    io::stdin().read_line(&mut input)?;
    let n: usize = input.trim().parse()?;
    
    let mut matriz: Matrix = Vec::with_capacity(n);
    println!("Digite a matriz aumentada (linha por linha):");

    for _ in 0..n {
        input.clear();
        io::stdin().read_line(&mut input)?;
        let linha: Vec<f64> = input
            .split_whitespace()
            .map(|x| x.parse().unwrap_or_else(|_| {
                eprintln!("Erro: Valor não numérico detectado");
                std::process::exit(1);
            }))
            .collect();
        
        if linha.len() != n + 1 {
            eprintln!("Erro: Número inválido de elementos na linha");
            std::process::exit(1);
        }
        matriz.push(linha);
    }

    let mut matriz_copy = matriz.clone();
    let solucao = eliminacao_gauss(&mut matriz_copy)?;

    println!("\nSolução:");
    for (i, valor) in solucao.iter().enumerate() {
        println!("x[{}] = {:.6}", i, valor);
    }

    Ok(())
}